server/services/prisonerInformation.js

```diff
@@ -0,0 +1,84 @@
+const { logger } = require('../utils/logger');
+
+function createPrisonMapFrom(prisonDetailsResponse) {
+  return prisonDetailsResponse
+    .filter(p => p !== null)
+    .reduce((a, p) => {
+      a.set(p.agencyId, p.longDescription);
+      return a;
+    }, new Map());
+}
+
+function formatTransactionsWith(transactions, prisonDetails) {
+  return transactions.map(t => ({
+    ...t,
+    prison: prisonDetails.has(t.agencyId)
+      ? prisonDetails.get(t.agencyId)
+      : t.agencyId,
+  }));
+}
+
+class PrisonerInformationService {
+  constructor({ prisonApiRepository }) {
+    this.prisonApi = prisonApiRepository;
+  }
+
+  async getTransactionInformationFor(user, accountCode, fromDate, toDate) {
+    if (!user || !accountCode || !fromDate || !toDate) {
+      throw new Error('Incorrect parameters passed');
+    }
+
+    try {
+      const [transactionsResponse, balancesResponse] = await Promise.all([
+        this.prisonApi.getTransactionsFor(
+          user.prisonerId,
+          accountCode,
+          fromDate,
+          toDate,
+        ),
+        this.prisonApi.getBalancesFor(user.bookingId),
+      ]);
+
+      const balances = balancesResponse || {
+        error: 'We are not able to show your balances at this time',
+      };
+
+      if (transactionsResponse) {
+        const listOfPrisons = Array.from(
+          new Set(transactionsResponse.map(t => t.agencyId)),
+        );
+        const prisonDetailsResponse = await Promise.all(
+          listOfPrisons.map(p => this.prisonApi.getPrisonDetailsFor(p)),
+        );
+
+        const prisonDetailsMap = createPrisonMapFrom(prisonDetailsResponse);
+        const formattedTransactions = formatTransactionsWith(
+          transactionsResponse,
+          prisonDetailsMap,
+        );
+
+        return {
+          transactions: formattedTransactions,
+          balances,
+        };
+      }
+
+      return {
+        transactions: {
+          error: 'We are not able to show your transactions at this time',
+        },
+        balances,
+      };
+    } catch (e) {
+      logger.error(
+        `PrisonerInformationService (getTransactionInformationFor) - Failed: ${e.message}`,
+      );
```

Comment:
You seem to have this handle error/ coerce to null pattern at multiple levels at both the api and service layer. 

In this module you aren't checking for the api returning nulls anywhere so presumably this catch would just be catching null pointer errors which seems like using exceptions for controlling flow.

Would probably be better to just throw exceptions from the api layer and catch here rather than generating multiple exceptions.

Also regarding the logging error - I think it's sometimes nice to visualise error logging as an event stream - ideally creating one "log error" event for each error that the system produces. 

Obviously this requires ensuring there is enough context in that one error event for you to be able to work out whats going on.

Reply:
In java/spring bootapps you'd generally just throw (or add context and rethrow) errors to be handled by a centralised error handler and allow that to do the logging. We've have some success with using an express error handler  for similar purposes

Reply:
My intent with swallowing the errors in the API tier was to allow for some graceful degradation when we unable to surface all the information for the page, such that if a call for their balance succeeds and the transactions call fails, then we would just show them their balance and notify them to try again for their transactions.

Interested in the logging by the way 🤔 would love to see some examples if you have them at hand